sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java

```diff
@@ -384,6 +399,72 @@ private PlannerResult planExplanation(
     return new PlannerResult(resultsSupplier, getExplainStructType(rel.getCluster().getTypeFactory()));
   }
 
+  /**
+   * Construct a {@link PlannerResult} for an 'explain' query from a {@link RelNode}
+   */
+  private PlannerResult planDruidExplanation(
+      final DruidRel<?> rel
+  )
+  {
+    String resources = null;
+    String explanation = null;
+    try {
+      explanation = dumpDruidPlan(rel);
+      resources = jsonMapper.writeValueAsString(plannerContext.getResources());
+    }
+    catch (JsonProcessingException jpe) {
+      // this should never happen, we create the Resources here, not a user
+      log.error(jpe, "Encountered exception while serializing Resources for explain output");
+    }
+    final Supplier<Sequence<Object[]>> resultsSupplier = Suppliers.ofInstance(
+        Sequences.simple(ImmutableList.of(new Object[]{explanation, resources})));
+    return new PlannerResult(resultsSupplier, getExplainStructType(rel.getCluster().getTypeFactory()));
+  }
+
+  /**
+   * This method doesn't utilize the Calcite's internal {@link RelOptUtil#dumpPlan} since that tends to be verbose
+   * and not indicative of the native Druid Queries which will get executed
+   * This method assumes that the Planner has converted the RelNodes to DruidRels, and thereby we can implictly cast it
+   *
+   * @param rel Instance of the root {@link DruidRel} which is formed by running the planner transformations on it
+   * @return A string representing an array of native queries that correspond to the given SQL query, in JSON format
+   * @throws JsonProcessingException
+   */
+  public static String dumpDruidPlan(DruidRel<?> rel) throws JsonProcessingException
+  {
+    List<Map<String, Object>> nativeQueries;
+
+    // Only if rel is an instance of DruidUnionRel, do we run multiple native queries corresponding to single SQL query
+    // Also, DruidUnionRel can only be a top level node, so we don't need to check for this condition in the subsequent
+    // child nodes
+    if (rel instanceof DruidUnionRel) {
+      nativeQueries = new ArrayList<>();
+      for (RelNode childRel : rel.getInputs()) {
+        DruidRel<?> druidChildRel = (DruidRel<?>) childRel;
+        if (childRel instanceof DruidUnionRel) { // Failsafe block, shouldn't be encountered
+          log.error("DruidUnionRel can only be the outermost Rel. This error shouldn't be encountered");
+          // TODO: Throw an error here
+        }
+        DruidQuery query1 = druidChildRel.toDruidQuery(false); // Finalize aggregations
+        nativeQueries.add(ImmutableMap.of(
+            "query",
+            query1.getQuery(),
+            "signature",
```

Comment:
I am not sure why we need the `signature` field. Is it also present in the existing EXPLAIN PLAN output?

Reply:
Yes, it is getting populated inside individual rel's `explainTerms` function. https://github.com/apache/druid/blob/223c5692a8292a210a770dc0702f75b4484ae347/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQueryRel.java#L199